Early publishing segments in the middle of data ingestion

jihoonson opened the pull request:
Fix #4227. Additionally, after this patch, IndexTask can be more improved by reading data from Firehose only once. I'll fix it in a follow-up pr.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/druid-io/druid/4238)
<!-- Reviewable:end -->

gianm commented:
@jihoonson, one thing to keep in mind is that users will probably want an option for whether to "guarantee" rollup or not. If it's guaranteed then the index task shouldn't publish early, and it probably needs to stick with two passes. If rollup is not guaranteed then we can take some short-cuts.

I haven't studied the patch yet so I'm not sure if this is addressed or not.

Sorry for commenting without reading but I just wanted to get that comment out :)

jihoonson commented:
@gianm thanks for your comment. I missed that point. I'll update this patch soon.

jihoonson commented:
@gianm btw, with this patch, the rollup is guaranteed when ```numShards``` is given. Is it able to be the way to specify guaranteed rollup you pointed out?

gianm commented:
It's still nice to have a way to guarantee rollup even when `numShards` is not provided.

jon-wei commented on indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java:
```diff
@@ -529,12 +494,55 @@ private FiniteAppenderatorDriver newDriver(
+     * @param row       input row
+     * @return a shardSpec
+     */
+    ShardSpec getShardSpec(Interval interval, long timestamp, InputRow row);
```
Do you foresee any situations where the timestamp here would be different from the row timestamp?

If not, maybe the timestamp parameter could be dropped and read from the row instead.

jihoonson replied:
Right. I removed the timestamp parameter.

jon-wei commented on server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java:
```diff
@@ -382,21 +432,30 @@ private SegmentIdentifier getSegment(final DateTime timestamp, final String sequ
+  private boolean moveSegmentOut(
+      final String sequenceName,
+      final List<SegmentIdentifier> identifiers,
+      boolean ignoreAbsent
```
Can you elaborate more on why ignoreAbsent is needed? It wasn't clear to me why this is needed at the call site with 'true'.

jihoonson replied:
When moving segments out after publishing them, some of the published segments may be already moved out from activeSegments due to the ```maxRowsPerSegment``` limit. I added a comment.

jon-wei commented on server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java:
```diff
@@ -178,47 +185,80 @@ public void clear() throws InterruptedException
 
   /**
-   * Add a row. Must not be called concurrently from multiple threads.
+   * Add a row.  This method may internally incur persisting data added so far.  Also, if too large data are persisted,
```
I wonder if any problems could arise from introducing this blocking segment publishing step to add() (uploading segments could take a while).

If so, would it be feasible/better to do the segment publishing asynchronously (maybe complicated since you could still run into disk space issues unless the add/drain rates are managed effectively, but would allow the input to continue to be processed while the segment push to deep storage is occurring).

jihoonson replied:
Currently, ```IndexTask``` is the only place where allows publishing segments in add(), and it is expected to use batch firehoses. So, I think usually there should be no issues. The one thing rose in my mind is that http connections can be expired if the underlying firehose is based on http and tries to keep the http connections during the whole ingestion step. But, this also can be avoided by using ```ReplayableFirehose``` or ```PrefetchableTextFilesFirehose``` introduced at #4193.

gianm replied:
@jihoonson what do you think about an alternate design:

- Change `add` to `AddResult add(InputRow row, String sequenceName, Supplier<Committer> committerSupplier)` where AddResult includes a SegmentIdentifier and something like numRowsAdded.
- Change `finish` to `publish` and have it return a `Future<SegmentsAndMetadata>` instead of `SegmentsAndMetadata`
- Make the task responsible for deciding when to call `publish` (instead of `add` like this patch has). The index task would make its decision based on numRowsAdded from AddResult.
- Rename the class from FiniteAppenderatorDriver to AppenderatorDriver. (It's no longer restricted to a finite stream!)

The idea behind this change is to make the AppenderatorDriver a good foundation for all kinds of incremental publishing, not just this one but also realtime incremental publishing like #4178. The Future return from publish is good because it prevents blocking realtime ingestion, and making the task responsible for decision about when to incrementally publish is good because the Kafka task will have some more complex logic around making sure all the replicas are in sync.

gianm replied:
@pjain1 @dclim does the above sound useful to you too, especially in the Kafka context? I'm wondering if it can reduce some of the complexity from #4178, maybe making DriverHolders unnecessary since a single AppenderatorDriver could handle multiple handoffs.

jihoonson replied:
@gianm @pjain1 @dclim sounds good to me, but i have a question.

> Change ```finish``` to ```publish``` and have it return a ```Future<SegmentsAndMetadata>``` instead of ```SegmentsAndMetadata```

here, which operation results in ```Future<SegmentsAndMetadata>```? More precisely, you mean the new ```publish()``` should return a future for segment handoff? or pushing segments? Currently, ```push()``` is an async operation which returns a future for pushing segments, while ```publish()``` is an atomic sync operation which writes metadata to the underlying dbms.

jihoonson commented on server/src/main/java/io/druid/segment/realtime/appenderator/Appenderator.java:
```diff
@@ -52,6 +52,8 @@
    */
   Object startJob();
 
+  long getPersistedBytes();
```
Added.

jihoonson commented on indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java:
```diff
@@ -222,30 +220,49 @@ public TaskStatus run(final TaskToolbox toolbox) throws Exception
    * hash-based partitioning). In the future we may want to also support single-dimension partitioning.
    */
-  private Map<Interval, List<ShardSpec>> determineShardSpecs(
+  private ShardSpecs determineShardSpecs(
```
Right, thanks. I'll update after addressing @gianm's [comment](https://github.com/druid-io/druid/pull/4238#issuecomment-298817242).

jihoonson replied:
Updated doc.

gianm commented on server/src/main/java/io/druid/segment/realtime/appenderator/Appenderator.java:
```diff
@@ -53,6 +53,14 @@
 
   /**
+   * Returns the size of data currently stored in local storage.  The persistedBytes is changed when
+   * {@link #persistAll(Committer)} and {@link #drop(SegmentIdentifier)} is called.
```
Not strictly true, since persistAll will return before persistedBytes is updated (the persisting happens in another thread). That might lead to persisting more data than we actually want to persist, since we're not counting either the following:

- data in memory in the currently active IncrementalIndex
- data in IncrementalIndexes that are pending persist in the persist thread

I'm not sure what's best to do here. Maybe we can track both rows and bytes and have a trigger to publish if either one exceeds a limit?

jihoonson replied:
I tried to trace the actual byte size of segments stored both in memory and local storage, but it looks to difficult to estimate the serialized byte size of not serialized objects without serialization.
So, I added another method ```getTotalRowCount()``` which returns the total number of rows in memory and disk instead of this one. 

Using row numbers as a kind of threshold is not good because operators should estimate proper values by themselves, so I think this should be improved to return number of bytes later.

jihoonson commented:
@gianm @jon-wei I've updated this patch against changes made in #4292. Also, I added an option ```forceGuaranteedRollup``` to ```IndexTuningConfig```.

jihoonson commented:
@gianm @jon-wei would you please review this patch again?

jon-wei commented on indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java:
```diff
@@ -431,41 +629,38 @@ public SegmentIdentifier allocate(DateTime timestamp, String sequenceName, Strin
-            if (!addResult.isOk()) {
-              throw new ISE("Could not allocate segment for row with timestamp[%s]", inputRow.getTimestamp());
+            if (addResult.isOk()) {
+              if (addResult.getNumRowsInSegment() >= maxRowsInSegment ||
```
Does this need a check to avoid the early publish if forceGuaranteedRollup=true?

jihoonson replied:
Nice catch! I added a check here and more tests for this.

jon-wei commented on docs/content/ingestion/tasks.md:
```diff
@@ -114,11 +114,13 @@ The tuningConfig is optional and default parameters will be used if no tuningCon
 |type|The task type, this should always be "index".|none|yes|
 |targetPartitionSize|Used in sharding. Determines how many rows are in each segment.|5000000|no|
 |maxRowsInMemory|Used in determining when intermediate persists to disk should occur.|75000|no|
+|maxTotalRows|Total number of rows in segments waiting for being published. Used in determining when intermediate publish should occur.|150000|no|
```
Can you add maybe another section to the docs about the task's publishing behavior? I think it'd be useful to have one place where the user can see what publishing modes are supported (early vs. no early) and what configuration parameters affect the publishing behavior

jihoonson replied:
Sure, I added more documents.

jihoonson commented on indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java:
```diff
@@ -214,59 +216,172 @@ public TaskStatus run(final TaskToolbox toolbox) throws Exception
-    final Map<Interval, List<ShardSpec>> shardSpecs = Maps.newHashMap();
+    final boolean determineIntervals = !granularitySpec.bucketIntervals().isPresent();
+    // Guaranteed rollup means that this index task guarantees the 'perfect rollup' across the entire data set.
+    final boolean guaranteedRollup = tuningConfig.isForceGuaranteedRollup() &&
```
Good catch. I changed to throw an exception.

jihoonson commented on docs/content/design/index.md:
```diff
@@ -56,30 +80,15 @@ This storage reduction does come at a cost; as we roll up data, we lose the abil
-## Sharding the Data
-
-Druid shards are called `segments` and Druid always first shards data by time. In our compacted data set, we can create two segments, one for each hour of data.
+### Roll-up modes
```
Thanks :)

jihoonson commented on docs/content/ingestion/tasks.md:
```diff
@@ -114,10 +114,12 @@ The tuningConfig is optional and default parameters will be used if no tuningCon
 |indexSpec|defines segment storage format options to be used at indexing time, see [IndexSpec](#indexspec)|null|no|
 |maxPendingPersists|Maximum number of persists that can be pending but not started. If this limit would be exceeded by a new intermediate persist, ingestion will block until the currently-running persist finishes. Maximum heap memory usage for indexing scales with maxRowsInMemory * (2 + maxPendingPersists).|0 (meaning one persist can be running concurrently with ingestion, and none can be queued  ...
 |forceExtendableShardSpecs|Forces use of extendable shardSpecs. Experimental feature intended for use with the [Kafka indexing service extension](../development/extensions-core/kafka-ingestion.html).|false|no|
+|forceGuaranteedRollup|Forces guaranteeing the [perfect rollup](./design/index.html). The perfect rollup optimizes the total size of generated segments and querying time while indexing time will be increased. This flag cannot be used with `appendToExisting` of IOConfig and `forceExtendableShardSpecs`. For more details, see the below __Segment publishing modes__ section.|false|no|
```
Changed to "This flag cannot be used with either `appendToExisting` of IOConfig or `forceExtendableShardSpecs`"

jihoonson commented on docs/content/ingestion/tasks.md:
```diff
@@ -148,6 +150,16 @@ For Roaring bitmaps:
+
+On the contrary, in the incremental publishing mode, segments are incrementally published, that is they can be published in the middle of the index task. More precisely, the index task collects data and stores created segments in the memory and disks of the node running that task until the total number of collected rows exceeds `maxTotalRows`. Once it exceeds, the index task immediately publishes ...
+
+To enable bulk publishing mode, `forceGuaranteedRollup` should be set in the TuningConfig. Note that this option cannot be used with `forceExtendableShardSpecs` of TuningConfig and `appendToExisting` of IOConfig together.
```
Changed.

jihoonson commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -213,11 +215,13 @@ public int add(
 
-    rowsCurrentlyInMemory.addAndGet(sinkRowsInMemoryAfterAdd - sinkRowsInMemoryBeforeAdd);
+    final int numAddedRows = sinkRowsInMemoryAfterAdd - sinkRowsInMemoryBeforeAdd;
+    rowsCurrentlyInMemory += numAddedRows;
```
Reverted and beefed up the contract.

jihoonson commented on indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java:
```diff
@@ -431,41 +638,40 @@ public SegmentIdentifier allocate(DateTime timestamp, String sequenceName, Strin
+                // Even though IndexTask uses NoopHandoffNotifier which does nothing for segment handoff,
+                // the below code is needed to update the total number of rows added to the appenderator so far.
+                // See AppenderatorDriver.registerHandoff() and Appenderator.drop().
+                // A hard-coded timeout is used here because the below get() is expected to return immediately.
```
Changed to 30 seconds

jihoonson commented on indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java:
```diff
@@ -577,10 +815,12 @@ public IndexIngestionSpec(
-                          : tuningConfig;
+      this.tuningConfig = tuningConfig == null ? new IndexTuningConfig() : tuningConfig;
+
+      if (this.ioConfig.isAppendToExisting() && this.tuningConfig.isForceGuaranteedRollup()) {
```
Ah, it looks fine to be removed.

jihoonson commented:
@gianm thanks. I addressed your comments.

gianm commented on server/src/main/java/io/druid/segment/realtime/appenderator/Appenderator.java:
```diff
@@ -61,8 +61,6 @@
    * Committer is guaranteed to be *created* synchronously with the call to add, but will actually be used
    * asynchronously.
-   * <p>
-   * The add, clear, persist, persistAll, and push methods should all be called from the same thread.
```
Hmm, I think these may still be needed. AppenderatorImpl's methods might not work properly if called concurrently from different threads. Also, nothing really needs to do that anyway. So I think we should leave in the restrictions on calling the major methods from the same thread.

jihoonson replied:
```add()```, ```persist()```, ```push()``` and ```drop()``` methods can be called by different threads in this patch. ```push()``` and ```persist()``` can be called by AppenderatorDriver.publishExecutor, and ```drop()``` is called by CoordinatorBasedSegmentHandoffNotifier.scheduledExecutor. Actually, ```persist()``` is called by AppenderatorImpl.pushExecutor even before this patch.

gianm replied:
Ah, I see! Ok, I will take one last look then.

gianm replied:
Is something making sure that "add" won't be called while these other methods are being called? What I'm concerned about is preventing a case where the Committer passed to "publish" is out of sync with what's being "add"ed.

I think if we need to relax the constraint that the methods are all called from the same thread, we should at least require they not be called from different threads at the same time. Maybe we can enforce this with synchronization? Otherwise, I'm concerned that commit metadata will get out of sync with the data itself.

Do you have any thoughts on this?

gianm replied:
Btw, the original idea was that the methods that accept Committer objects should all be called from the same thread, because that ensures that the committed metadata on disk is in sync with the data itself.

jihoonson replied:
@gianm Hmm, I missed it. I think synchronization is not enough because we need to guarantee that the execution order matches the method call order. Otherwise, we need a kind of sorted queue which provides such a guarantee. I think the easiest way is enforcing the methods are called by the same thread as before. 

One thing I'm concerned with is changing ```Appenderator.push()``` looks inevitable to implement the same retry mechanism of ```AppenderatorDriver.publish()```. All ```Appenderator.add()```, ```Appenderator.push()```, and ```Appenderator.persist()``` methods should be called by the same thread calling ```AppenderatorDriver```, and to guarantee this in ```AppenderatorDriver.publish()```, ```Appenderator.push()``` should provide the retry mechanism which retries the push until complete because the calling thread of ```push()``` should be ```pushExecutor```. I think it makes sense to add a new push() method which takes a max retry count. 

What do you think?

gianm replied:
> I think synchronization is not enough because we need to guarantee that the execution order matches the method call order. Otherwise, we need a kind of sorted queue which provides such a guarantee. I think the easiest way is enforcing the methods are called by the same thread as before.

Yeah, I think that if it's feasible to require that the methods are called by the same thread, then that's simplest from the point of view of the Appenderator implementation.

> I think it makes sense to add a new push() method which takes a max retry count.

You mean `Appenderator.push()`, and having that method do its own retries? That makes sense to me. The ability to do retries in the Appenderator certainly makes life easier for its callers.

It sounds like it also makes it easier to ensure that the Committer metadata is always in sync with data written to disk, which is essential for correct behavior.

jihoonson replied:
Thanks. I fixed the sync problem in committed metadata. 

Also, I changed AppenderatorDriver.publish() to not retry. Instead, Appenderator does retry itself only for pushing segments. This is because exceptions from other parts are usually IOException while persisting segments or NPE due to logic error, etc and I don't think these exceptions are recoverable by executing multiple times.

jihoonson commented:
Fixed conflicts

jihoonson commented:
@gianm would you review again the latest patch?

gianm commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -576,9 +595,14 @@ private DataSegment mergeAndPush(final SegmentIdentifier identifier, final Sink
-          mergedFile,
-          sink.getSegment().withDimensions(IndexMerger.getMergedDimensionsFromQueryableIndexes(indexes))
+      // Retry pushing segments because uploading to deep storage might fail especially for cloud storage types
+      final DataSegment segment = RetryUtils.retry(
```
I wonder if the retrying the push should be the responsibility of the pusher, rather than the Appenderator. S3DataPusher already does its own retries, so the retries here are not really helping anything for S3. The pushers are also in a position to potentially be smarter about retries (the S3 one won't retry "access denied" for example; see `S3Utils.retryS3Operation`; so it can fail faster when retrying would be useless).

What do you think?

jihoonson replied:
It makes sense and sounds even better. I think its better to make the pushers to retry by themselves in a follow-up pr because there are somethings to do like checking the ```S3RETRY``` predicate can be applied to other types of pushers. Maybe it's possible to make the retrying smarter. What do you think?

gianm replied:
Sure, I think it's fine to do it in another pr.

jihoonson replied:
Thanks! I raised #4529.

gianm commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -307,20 +319,22 @@ public void clear() throws InterruptedException
-          {
-            @Override
-            public Object call() throws Exception
+      if (persistExecutor != null) {
```
Oops, meant to say I don't think it can be _null_. Edited the comment.

jihoonson replied:
I saw a test failure due to NPE from here, but can't find what it was.. Maybe there were some bugs which are fixed now.

jihoonson commented on server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -352,10 +366,15 @@ public Object call() throws Exception
     for (SegmentIdentifier identifier : identifiers) {
       final Sink sink = sinks.get(identifier);
+      if (sink == null) {
+        throw new NullPointerException("No sink for identifier: " + identifier);
```
Fixed.

jihoonson commented:
@gianm thank you for the review!

leventov commented on indexing-service/src/main/java/io/druid/indexing/common/task/IndexTask.java:
```diff
@@ -214,61 +216,182 @@ public TaskStatus run(final TaskToolbox toolbox) throws Exception
+    final Map<Interval, Optional<HyperLogLogCollector>> hllCollectors = new TreeMap<>(
+        Comparators.intervalsByStartThenEnd()
+    );
     int thrownAway = 0;
```
This variable is never updated. The if below conditioning of this variable to be positive is dead code. See #7737.
